### PR TITLE
Wait for cluster to recover before resolving index template

### LINKED ISSUE #49499: Index template resolution does not wait for the Cluster State to be recovered
If a bulk request hits a node that has not recovered the CS yet, index metadata will not be available. The actual bulk request will wait/retry until the CS is recovered (due to retry mechanism in BulkOperation), but the resolution of index templates has no retry mechanism. This means default-pipeline processors will be omitted and this can result in incorrect documents being indexed.  It may affect other aspects that deal with index templates as well.

There was an attempt to fix this (https://github.com/elastic/elasticsearch/pull/46085), but the PR stalled.  I closed that PR so that the issue is "unclaimed" in case someone else wants to fix the issue.  I might revisit, but don't have time right now. :(

### PR DESCRIPTION
**Pull request by ywangd**
This PR makes bulk index action to wait for cluster to recover before resolving index templates so that ingest pipelines are correctly processed when the cluster is recovering.

Resolves: #49499
Supersedes: #46085

### CODE REVIEW

File: modules/ingest-common/src/internalClusterTest/java/org/elasticsearch/ingest/common/IngestRestartIT.java [hunk 1]
```diff
@@ -262,4 +272,95 @@ public void testWithDedicatedIngestNode() throws Exception {
...
+                @Override
+                public void onFailure(Exception e) {
+                    fail("Should not have failed with exception: " + e.getMessage());
+                }
+            });
+
+        // so the cluster state can be recovered
+        internalCluster().startNode(Settings.builder().put(GatewayService.RECOVER_AFTER_DATA_NODES_SETTING.getKey(), "1"));
+        ensureYellow("index");
+        assertTrue(latch.await(5, TimeUnit.SECONDS));
```
Comment (DaveCTurner) [inline modules/ingest-common/src/internalClusterTest/java/org/elasticsearch/ingest/common/IngestRestartIT.java:348]: 5s might be a little short, suggest using `safeAwait`:

```suggestion
        safeAwait(latch);
```

However, personally I think it'd be even better to use a `PlainActionFuture` as the listener so that any exception is immediately thrown back on the main test thread rather than having to wait for the timeout on the latch.

File: modules/ingest-common/src/internalClusterTest/java/org/elasticsearch/ingest/common/IngestRestartIT.java [hunk 2]
```diff
@@ -262,4 +272,95 @@ public void testWithDedicatedIngestNode() throws Exception {
...
+
+        client().prepareIndex("index").setId("passes2").setSource("x", 3).setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE).get();
+        client().admin().indices().prepareRefresh("index").get();
```
Comment (DaveCTurner) [inline modules/ingest-common/src/internalClusterTest/java/org/elasticsearch/ingest/common/IngestRestartIT.java:351]: This refresh shouldn't be necessary I think, we set `?refresh=true` on all the preceding indexing.

File: modules/ingest-common/src/internalClusterTest/java/org/elasticsearch/ingest/common/IngestRestartIT.java [hunk 3]
```diff
@@ -262,4 +272,95 @@ public void testWithDedicatedIngestNode() throws Exception {
...
+        Map<String, Object> source = client().prepareGet("index", "passes1").get().getSource();
+        assertThat(source.get("x"), equalTo(2));
+        assertThat(source.get("value"), equalTo(42));
+
+        source = client().prepareGet("index", "passes2").get().getSource();
+        assertThat(source.get("x"), equalTo(3));
+        assertThat(source.get("value"), equalTo(42));
+
+        // and make sure this failed doc didn't get through
+        source = client().prepareGet("index", "fails").get().getSource();
```
Comment (DaveCTurner) [inline modules/ingest-common/src/internalClusterTest/java/org/elasticsearch/ingest/common/IngestRestartIT.java:363]: I would prefer all of these `.get()` calls to have a timeout (10s?) too.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 4]
```diff
@@ -191,9 +191,47 @@ public static <Response extends ReplicationResponse & WriteResponse> ActionListe
...
+                public void onTimeout(TimeValue timeout) {
+                    listener.onFailure(blockException);
+                }
+            }, newState -> false == newState.blocks().hasGlobalBlockWithLevel(ClusterBlockLevel.WRITE));
+        } else {
+            doExecuteOnWriteThreadPool(task, bulkRequest, startTime, listener);
+        }
+    }
+
+    private void doExecuteOnWriteThreadPool(Task task, BulkRequest bulkRequest, long startTime, ActionListener<BulkResponse> listener) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:231]: Naming nit: I interpreted this name as indicating that it's _called_ on the write threadpool. Maybe something like `forkAndExecute` would be clearer?

```suggestion
    private void forkAndExecute(Task task, BulkRequest bulkRequest, long startTime, ActionListener<BulkResponse> listener) {
```

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 5]
```diff
@@ -209,19 +247,24 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
...
             @Override
             protected void doRun() {
-                doInternalExecute(task, bulkRequest, executorName, releasingListener);
+                doInternalExecute(task, bulkRequest, executorName, startTime, releasingListener);
             }
         });
     }
 
-    protected void doInternalExecute(Task task, BulkRequest bulkRequest, String executorName, ActionListener<BulkResponse> listener) {
-        final long startTime = relativeTime();
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:224]: I think preserving the original start time is a good change but it seems unrelated to the rest of this PR, and deserves a test of its own too. Could you separate that out into a different PR?

**Review by DaveCTurner — COMMENTED** [review]
A blast from the past, thanks for picking this up Yang. Looks good, I left only superficial comments.

File: modules/ingest-common/src/internalClusterTest/java/org/elasticsearch/ingest/common/IngestRestartIT.java [hunk 6]
```diff
@@ -262,4 +272,95 @@ public void testWithDedicatedIngestNode() throws Exception {
...
+
+        internalCluster().fullRestart(new InternalTestCluster.RestartCallback() {
+            @Override
+            public Settings onNodeStopped(String nodeName) {
+                return Settings.builder().put(GatewayService.RECOVER_AFTER_DATA_NODES_SETTING.getKey(), "2").build();
+            }
+
+            @Override
+            public boolean validateClusterForming() {
+                return false;
```
[the remark this answers is not in the document]
Comment (ywangd) [reply modules/ingest-common/src/internalClusterTest/java/org/elasticsearch/ingest/common/IngestRestartIT.java]: That's good know. Thanks. Updated as suggested.

File: modules/ingest-common/src/internalClusterTest/java/org/elasticsearch/ingest/common/IngestRestartIT.java (hunk 3 above)
Comment (ywangd) [reply modules/ingest-common/src/internalClusterTest/java/org/elasticsearch/ingest/common/IngestRestartIT.java]: 👍  Updated all the places with a 10 second timeout.

File: modules/ingest-common/src/internalClusterTest/java/org/elasticsearch/ingest/common/IngestRestartIT.java (hunk 1 above)
Comment (ywangd) [reply modules/ingest-common/src/internalClusterTest/java/org/elasticsearch/ingest/common/IngestRestartIT.java]: I agree. Using `PlainActionFuture` is a better and more explicit choice. Updated.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java (hunk 5 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: Thanks for catching it. This is a bigger change than I initially thought because today the startTime is reset after ingestService completes the processing. My previous change would have made it include the ingestService processing time (but not the request forward time). I have reverted this change. A follow-up is better. Thanks!

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 7]
```diff
@@ -209,9 +246,9 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
         final int indexingOps = bulkRequest.numberOfActions();
         final long indexingBytes = bulkRequest.ramBytesUsed();
         final boolean isOnlySystem = isOnlySystem(bulkRequest, clusterService.state().metadata().getIndicesLookup(), systemIndices);
+        final String executorName = isOnlySystem ? Names.SYSTEM_WRITE : Names.WRITE;
         final Releasable releasable = indexingPressure.markCoordinatingOperationStarted(indexingOps, indexingBytes, isOnlySystem);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:250]: Ah wait, sorry, we need to do this check first, on the transport worker, before we start waiting for cluster states. Otherwise we might just pile up far too much work in memory waiting for the cluster recovery.

But that is a little tricky because we need to compute `isOnlySystem` to call this, and that needs the cluster to be recovered. I suggest we conservatively assume `isOnlySystem == false` if the cluster is not yet recovered.

Comment (ywangd) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:250]: Sorry I don't quite follow. It seems to me that when the code reaches here, the cluster state is either recovered successfully or we had no need to wait for recover at all. So there is no more waiting after this line?

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:250]: Yes that's right, but it's the waiting _before_ this line that's a problem. Before we run this line we have the bulk request in memory but we aren't tracking it in the indexing pressure subsystem. If we receive lots of indexing requests while the cluster is recovering then we will try to hold them all in memory, rejecting none of them and will eventually just OOM.

Comment (ywangd) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: Makes sense. Thanks a lot for the explanation! Updated in 812154a8acf646b4c722e171d6200420838297af

Comment (ywangd) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: Btw, I didn't change how `isOnlySystem` is computed since the method is written with the assumption that the index may not be available yet (see [here](https://github.com/elastic/elasticsearch/pull/61655/files#r491024993)).
Also, if the metadata is not available, the ultimate default is to return `false` (when the `SystemIndices` also does not know the index).

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 8]
```diff
@@ -206,18 +206,58 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
...
-        final Releasable releasable = indexingPressure.markCoordinatingOperationStarted(indexingOps, indexingBytes, isOnlySystem);
-        final ActionListener<BulkResponse> releasingListener = ActionListener.runBefore(listener, releasable::close);
+        final boolean isOnlySystem = isOnlySystem(bulkRequest, initialState.metadata().getIndicesLookup(), systemIndices);
         final String executorName = isOnlySystem ? Names.SYSTEM_WRITE : Names.WRITE;
-        threadPool.executor(Names.WRITE).execute(new ActionRunnable<>(releasingListener) {
-            @Override
-            protected void doRun() {
-                doInternalExecute(task, bulkRequest, executorName, releasingListener);
+        final Releasable releasable = indexingPressure.markCoordinatingOperationStarted(indexingOps, indexingBytes, isOnlySystem);
+        // We should use the releasingListener from here onwards
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:215]: A nice way to enforce this is to extract the rest of this method out into another method which receives only the wrapped listener in its arguments.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 9]
```diff
@@ -206,18 +206,58 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
...
+                    releasingListener.onFailure(blockException);
+                }
+            }, newState -> false == newState.blocks().hasGlobalBlockWithLevel(ClusterBlockLevel.WRITE));
+        } else {
+            threadPool.executor(Names.WRITE).execute(new ActionRunnable<>(releasingListener) {
+                @Override
+                protected void doRun() {
+                    doInternalExecute(task, bulkRequest, executorName, releasingListener);
+                }
+            });
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:259]: This is the same as the `onNewClusterState` implementation, can we find a way to reduce the duplication?

Also I wonder why we're forking to `WRITE` here even though we know it might be a low-latency system-indices-only write. Not a new problem, but something to think about in future.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 10]
```diff
@@ -206,18 +206,58 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
...
-        final ActionListener<BulkResponse> releasingListener = ActionListener.runBefore(listener, releasable::close);
+        final boolean isOnlySystem = isOnlySystem(bulkRequest, initialState.metadata().getIndicesLookup(), systemIndices);
         final String executorName = isOnlySystem ? Names.SYSTEM_WRITE : Names.WRITE;
-        threadPool.executor(Names.WRITE).execute(new ActionRunnable<>(releasingListener) {
-            @Override
-            protected void doRun() {
-                doInternalExecute(task, bulkRequest, executorName, releasingListener);
+        final Releasable releasable = indexingPressure.markCoordinatingOperationStarted(indexingOps, indexingBytes, isOnlySystem);
+        // We should use the releasingListener from here onwards
+        final ActionListener<BulkResponse> releasingListener = ActionListener.runBefore(originalListener, releasable::close);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:216]: I have a slight preference for leaving these lines in the same order to simplify the diff, unless I'm missing something about the new order being required?

**Review by DaveCTurner — APPROVED** [review]
LGTM as it stands although I left a few more optional suggestions too.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java (hunk 7 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: I had some concerns about whether `clusterState.metadata()` could be null, or `clusterState.metadata().getIndicesLookup()`, but I did some checking and I think this is ok.

Comment (ywangd) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: Thanks checking! I didn't think that was possible.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java (hunk 8 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: Yeah I thought about that, but somehow decided against for no particularly good reason. I have now split it into a separate method.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java (hunk 9 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: It is now splitted into a separate method. I didn't do that because it felt too small for a method and not too big for a duplication. It is also possible to wrap the thing in a runnable so that it can be called in both places. But it feels wasteful to create an extra object. Please let me know if you have a preference.

 > Also I wonder why we're forking to WRITE here even though we know it might be a low-latency system-indices-only write. Not a new problem, but something to think about in future.

Yeah. I noticed it as well when I duplicated the code. I think it is a bug. But would rather address it separately.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java (hunk 10 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: It was accidental. Should be reverted by now.

**Review by DaveCTurner — APPROVED** [review]
Looks great, thanks for the extra iterations
